src/estimators/same_key_aggregator.py

```diff
@@ -94,54 +98,107 @@ def assert_compatible(self, other):
+        if frequency_noise_class guarantees purely epsilon-DP.
+      reach_noiser_kwargs: a dictionary of other kwargs to be specified in
+        reach_noiser_class. For example, it can be an empty dictionary, or
+        {'random_state': <a np.random.RandomState>}.
+      frequency_noiser_kwargs: a dictionary of other kwargs to be specified in
+        frequency_noiser_class. For example, it can be an empty dictionary, or
+        {'random_state': <a np.random.RandomState>}.
     """
     self.max_freq = max_freq
-    self.one_plus_reach_noiser = None
-    self.histogram_noiser = None
-    if noiser_class is not None:
-      assert epsilon_split > 0 and epsilon_split < 1, (
-          'In StandardizedHistogramEstimator, epsilon_split must be >0 and <1.')
-      # self.one_plus_reach_noiser is the noiser on the estimate of 1+ reach.
-      self.one_plus_reach_noiser = noiser_class(epsilon=epsilon * epsilon_split)
-      # self.histogram_noiser is the noiser on the frequency histogram among
-      # effective keys
-      # For simplicity, suppose the two noisers share the same noiser_class.
-      self.histogram_noiser = noiser_class(
-          epsilon=epsilon * (1 - epsilon_split))
+    self.one_plus_reach_noiser, self.reach_epsilon, self.reach_delta = (
+        StandardizedHistogramEstimator.define_noiser(
+            reach_noiser_class, reach_epsilon, reach_delta, reach_noiser_kwargs))
+    self.histogram_noiser, self.frequency_epsilon, self.frequency_delta = (
+        StandardizedHistogramEstimator.define_noiser(
+            frequency_noiser_class, frequency_epsilon,
+            frequency_delta, frequency_noiser_kwargs))
+    self.total_epsilon, self.total_delta = self.calculate_total_privacy_budget()
 
   def __call__(self, sketch_list):
     ska = StandardizedHistogramEstimator.merge_sketch_list(sketch_list)
     return self.estimate_cardinality(ska)
 
+  @classmethod
+  def define_noiser(cls, noiser_class, epsilon, delta=0, kwargs={}):
+    """A util for defining reach_noiser and frequency_noiser.
+
+    Args:
+      noiser_class: given name of noiser_class.
+      epsilon: given epsilon for the noiser.
+      delta: given delta for the noiser.
+      kwargs: other args to be specified in the noiser_class.
+
+    Returns:
+      a tuple of (<obtained noiser>, <actual epsilon the noiser guarantees>,
+      <actual delta the noiser guarantees>)
+    """
+    if noiser_class is None:
+      return (None, np.Inf, 0)
+    if delta in inspect.getfullargspec(noiser_class)[0]:
+      # If delta is an argument of noiser_class
+      return (noiser_class(epsilon=epsilon, delta=delta, **kwargs),
+              epsilon, delta)
+    # Otherwise, delta is not an argument of noiser_class
+    warnings.warn('The given delta is ignored, since it is not in the noiser.')
+    return (noiser_class(epsilon=epsilon, **kwargs), epsilon, 0)
+
+  def calculate_total_privacy_budget(self):
```

Comment:
Rather than calculating the total privacy consumption and returning it, I think it would be better to return a list of (epsilon, delta) pairs representing the noise amounts that were added to the data for the different noising events.  These different noising events could then be individually recorded in the PrivacyTracker object in the planning eval framework.  This would facilitate using the advanced composition theorem to calculate the total privacy usage over the run of a modeling strategy (or multiple runs of modeling strategies).  Also, it would centralize the computation of privacy consumption in a single location in the code.  So, if we later decide to add support for PLD, only one place in the code will have to be changed.  What do you think?

Reply:
+1 regarding noising events as those might allow us to do tighter compositions